Update Contributor Guidelines (#82)
Gilead-BioStats/gsm.core: vignettes/ContributorGuidelines.Rmd

* Always **link PRs to their issues** using [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue-using-a-keyword).
* PRs must pass all **CI checks** (tests, style, docs, coverage) before merge.
* Default target branch is `dev`; only **release** PRs target `main`.

👉 A well-prepared PR makes reviewing faster and ensures that releases are low-risk.

---

## Release Workflow

Releases move code from development into production. The goal is to ensure **reliability, traceability, and reproducibility**:

1. **Create a `release-x.y.z` branch** from `dev` (using [semantic versioning](https://semver.org/)).
2. **Prepare the release** – update version, NEWS, documentation, qualification files, and run checks (`devtools::check()`, spell check, `pkgdown::build_site()`).
3. **Open a release PR** into `main` using the release template. Assign yourself and request QC review.
4. **QC reviewers** complete checklists and may request fixes. Any fixes are made in `fix-*` branches and merged back into the release branch.
5. **Complete the release** – once approved, follow the following steps:
    * Merge PR into `main`

>     * Create a GitHub Release

Reviewer: Are there any specific requirements for the GitHub release to be followed?

# Project Management

The project management workflow focuses on using issues to capture user requirements, bugs, and technical requirements. We also use the [gsm Roadmap](https://github.com/orgs/Gilead-BioStats/projects/41) GitHub Project to track issues and set priorities across repositories. 

## Issues

Issues are the primary way to communicate what needs to be done and to track progress. Several issue templates are provided to help streamline this process, including: 

- [Requirements](https://github.com/Gilead-BioStats/gsm.core/issues/new?template=business_requirement.md): Use this template to create a User Requirement. Requirements are then assigned sub-issues using the issue types below. 
- [Bug](https://github.com/Gilead-BioStats/gsm.core/issues/new?template=bug_report.md): Bug reports for when something isn't working
- [Feature](https://github.com/Gilead-BioStats/gsm.core/issues/new?template=feature_request.md): User-facing Functionality
- [Technical Task](https://github.com/Gilead-BioStats/gsm.core/issues/new?template=qc_update.md): Non-user facing tasks, such as infrastructure updates or internal tooling improvements.

> - [Documentation Task](https://github.com/Gilead-BioStats/gsm.core/issues/new?template=documentation.md): Improvements or additions to the documentation including function docs, readme updates and vignettes.

Reviewer 2: These are great here, but we might want to link directly to the templates instead of the "new issue" page, so it doesn't behave surprisingly when we link from other repos. Alternatively, leave this as-is, and we'll solve it in a separate PR for the other repos, but I *think* you want to nail this down.

> The issue templates are maintained in {[gsm.utils](https://github.com/Gilead-BioStats/gsm.utils)} and automatically updated in all gsm packages whenever updates are made. Note that suggestions or other input that might not warrant formal submission of an issue can be filed under [discussions](https://github.com/Gilead-BioStats/gsm.core/discussions), which can help facilitate discourse of specific use-cases or requests.

Reviewer 2: The discussions link might also need to be generalized. Or we can have a central discussion board if that makes sense, but I imagine it would likely be better to have package-specific discussions in their repos.

Reviewer 3: I don't think we _need_ the link here. It's relatively easy to locate the Discussions tab in the github repo, and would simplify this.

---

## Road Map

> The [gsm Roadmap](https://github.com/orgs/Gilead-BioStats/projects/41) GitHub Project is used to track issues and set priorities across projects. The gsm team generally batches work in to quarterly releases, although package releases can happen more frequently. Key cross-repo views in the project include: 

Reviewer 2: ```suggestion
The [gsm Roadmap](https://github.com/orgs/Gilead-BioStats/projects/41) GitHub Project is used to track issues and set priorities across projects. The gsm team generally batches work into quarterly releases, although package releases can happen more frequently. Key cross-repo views in the project include: 
```

- Roadmap - a list of issues split out by quarter. Generally speaking, related functionality should be grouped in to a Requirement. Key columns include: 
   - Status - Issue status (e.g. To Do, In Progress, Done)
   - Repository/Milestone
   - Roadmap - Quarter where the requirement is planned for completion
   - Triage - has the requirement been approved for development work
   - Type - Bug, Feature, Task, Documentation
- Issues - a list of all open non-requirement issues, split by type.

> - Releases - All open issues grouped by repo and sorted by milestone. 

Reviewer 2: ```suggestion
- Roadmap – A list of issues split out by quarter. Generally speaking, related functionality should be grouped into a Requirement. Key columns include: 
   - Status – Issue status (e.g. To Do, In Progress, Done).
   - Repository/Milestone
   - Roadmap – Quarter when the requirement is planned for completion.
   - Triage – Has the requirement been approved for development work?
   - Type – Bug, Feature, Task, Documentation.
- Issues – A list of all open non-requirement issues, split by type.
- Releases – All open issues grouped by repo and sorted by milestone. 
```
Consistency.

# Development Workflow

Development is done using R package development best practices ([R Packages](https://r-pkgs.org/)) and the **GitHub Flow** workflow ([GitHub Flow Guide](https://docs.github.com/en/get-started/using-github/github-flow)).
The goal is to keep code reliable, traceable, and easy for everyone to contribute.

---

## Branches

We use a simple branching model to keep work organized:

* **`main`** – stable, production-ready code. Every package release comes from here.
* **`dev`** – working branch that collects all new features, bugfixes, and QC tasks before release.
* **`fix-*`** – short-lived branches for specific issues (feature, bug, or QC task). All development starts here.

> * **`release-*`** – temporary branches for regression testing, QC, and documentation before merging into `main`.

Reviewer 2: ```suggestion
* **`main`** – Stable, production-ready code. Every package release comes from here.
* **`dev`** – Working branch that collects all new features, bugfixes, and QC tasks before release.
* **`fix-*`** – Short-lived branches for specific issues (feature, bug, or QC task). All development starts here.
* **`release-*`** – Temporary branches for regression testing, QC, and documentation before merging into `main`.
```

👉 This structure ensures that unfinished work never goes directly into production, but can still be tested and reviewed in isolation.

---

## Development Workflow

> 1. **Open an issue** – every change begins with an issue using the right template (`Bug`, `Feature`,`Technical Task`, or `Documentation Task`). This creates a clear record of what’s being worked on.

Reviewer: ```suggestion
1. **Open an issue** – every change begins with an issue using the right template (`Requirements`, `Bug`, `Feature`,`Technical Task`, or `Documentation Task`). This creates a clear record of what’s being worked on.
```

2. **Create a branch** – branch names link to issues (e.g., `fix-111` for issue #111) so progress is traceable.
3. **Develop & document** – write clean, consistent code, add roxygen2 documentation, and create/update [testthat](https://testthat.r-lib.org/) unit tests.
4. **Open a Pull Request** – submit your branch to `dev`, assign yourself, request a reviewer, and link the issue. CI checks will run automatically.

> 5. **Merge & clean up** – once approved and passing checks, merge into `dev` and delete the `fix-*` branch.

Reviewer 2: ```suggestion
1. **Open an issue** – Every change begins with an issue using the right template (`Bug`, `Feature`,`Technical Task`, or `Documentation Task`). This creates a clear record of what’s being worked on.
2. **Create a branch** – Branch names link to issues (e.g., `fix-111` for issue #111) so progress is traceable.
3. **Develop & document** – Write clean, consistent code, add roxygen2 documentation, and create/update [testthat](https://testthat.r-lib.org/) unit tests.
4. **Open a Pull Request** – Submit your branch to `dev`, assign yourself, request a reviewer, and link the issue. CI checks will run automatically.
5. **Merge & clean up** – Once approved and passing checks, merge into `dev` and delete the `fix-*` branch.
```

👉 Following this flow ensures that every contribution is tracked, tested, and reviewed before becoming part of the development branch.

---

## Pull Requests

Pull Requests are where code review happens. To keep them smooth:

> * Follow the [tidyverse style guide](https://style.tidyverse.org/) See the Code Style section below for details on styling code per GSM standards.

Reviewer 2: ```suggestion
* Follow the [tidyverse style guide](https://style.tidyverse.org/). See the Code Style section below for details on styling code per GSM standards.
```

* Default target branch is `dev`; only **release** PRs target `main`.

👉 A well-prepared PR makes reviewing faster and ensures that releases are low-risk.

---

## Release Workflow

Releases move code from development into production. The goal is to ensure **reliability, traceability, and reproducibility**:

1. **Create a `release-x.y.z` branch** from `dev` (using [semantic versioning](https://semver.org/)).
2. **Prepare the release** – update version, NEWS, documentation, qualification files, and run checks (`devtools::check()`, spell check, `pkgdown::build_site()`).
3. **Open a release PR** into `main` using the release template. Assign yourself and request QC review.
4. **Link PR to relevant parent issue** in the [gsm Roadmap project board](https://github.com/orgs/Gilead-BioStats/projects/41/). 
5. **QC reviewers** complete checklists and may request fixes. Any fixes are made in `fix-*` branches and merged back into the release branch.
6. **Complete the release** – once approved, follow the following steps:
    * Merge PR into `main`
    * Create a GitHub Release
    * Attach QC reports
    * Create a PR that syncs `main` back into `dev`

>     * Ensure related Issues are Closed

Reviewer 2: ```suggestion
2. **Prepare the release** – Update version, NEWS, documentation, qualification files, and run checks (`devtools::check()`, spell check, `pkgdown::build_site()`).
3. **Open a release PR** into `main` using the release template. Assign yourself and request QC review.
4. **Link PR to relevant parent issue** in the [gsm Roadmap project board](https://github.com/orgs/Gilead-BioStats/projects/41/). 
5. **QC reviewers** complete checklists and may request fixes. Any fixes are made in `fix-*` branches and merged back into the release branch.
6. **Complete the release** – Once approved, follow the following steps:
    * Merge PR into `main`.
    * Create a GitHub Release.
    * Attach QC reports.
    * Create a PR that syncs `main` back into `dev`.
    * Ensure related Issues are Closed.
```

👉 This process makes sure that every release is tested, documented, and signed off before it goes live.

---

## Code Style

We follow the [tidyverse style guide](https://style.tidyverse.org/) with minor tweaks.
Before each release, code is standardized with `styler` so formatting is consistent across the whole repo.

> ```{r eval = FALSE}

Reviewer 2: ````suggestion
```r
````
This is what this becomes when rendered.

double_indent_style <- styler::tidyverse_style()
double_indent_style$indention$unindent_fun_dec <- NULL
double_indent_style$indention$update_indention_ref_fun_dec <- NULL
double_indent_style$line_break$remove_line_breaks_in_fun_dec <- NULL
styler::style_dir("R", transformers = double_indent_style)
styler::style_dir("tests", recursive = TRUE, transformers = double_indent_style)
```

---

## Continuous Integration

GitHub Actions run automatically on all Pull Requests to `dev` and `main` to catch issues early:

> * **R CMD check** – confirms the package builds and passes tests.
> * **Pkgdown** – Generates `pkgdown` docs.
> * **Release checks** – run on multiple platforms (Linux, macOS, Windows) to confirm reproducibility.
> * **Qualification Reports** – built and attached to releases for traceability. *Only run on merges to `main`.*

Reviewer 2: ```suggestion
* **R CMD check** – Confirms the package builds and passes tests.
* **Pkgdown** – Generates `pkgdown` docs.
* **Release checks** – Run on multiple platforms (Linux, macOS, Windows) to confirm reproducibility.
* **Qualification Reports** – Built and attached to releases for traceability. *Only run on merges to `main`.*
```

> 

Reviewer: ```
**Release** - Builds and attaches tarball to releases

```

👉 Contributors don’t typically need to configure anything—just push code and the checks will run.

---

# Appendix 1 – Quick Reference

### Fix Branch Workflow

1. Open issue with correct template.
2. Create `fix-*` branch.
3. Develop code + docs + tests.
4. Open PR to `dev` (assign self, request review, link issue).
5. Merge after approval + passing checks.

### Release Branch Workflow

1. Create `release-x.y.z` branch from `dev`.
2. Update version, NEWS, docs, qualification files.
3. Run checks locally + CI in a clean working environment.
4. Open release PR to `main`.
5. QC reviewers complete checklist.
6. Merge, publish GitHub Release, attach QC report.

> 7. Sync `main` → `dev`, close issues, clean up branches.

Reviewer: Do we open a PR from `main` to `dev` or syncing refers to some other action in this case?

Reviewer 3: added language to make this more clear

### QC Checklist

* [ ] Roxygen2 docs complete (`@export`, `@param`, `@return`, examples).
* [ ] Unit tests for inputs, outputs, error handling.

> * [ ] `devtools::check()` passes (no errors/warnings/notes).

Reviewer: Should programmer run `devtools::check()` locally on the branch to confirm?

Reviewer 3: right. do i need to make this more clear- something like "Run `devtools::check() and ensure checks pass locally (no errors/warnings/notes).

Reviewer: I think adding more details won't hurt. Thank you!

* [ ] No sensitive data or hardcoded paths.
* [ ] Qualification specs + reports updated.
* [ ] All GitHub Actions checks pass.

---

# Appendix 2- Quarterly Release Instructions
  
This document describes the process for performing a quarterly release of the `{gsm}` suite of R packages on GitHub. The primary difference between this release and off-cycle releases is that given the likelihood of multiple packages being simultaneously updated, the releases must be done in a specified order to avoid dependency conflicts.

---
  
## Release Workflow for Quarterly Releases

This workflow is similar to the standard release workflow, but with additional emphasis on ensuring dependencies are up to date and properly listed in `DESCRIPTION` given this quarter's release plan.
  

> 1. Create a **release branch** (e.g., `release-QX.Y`) from `main` in each package repository.

Reviewer: `release-Q.X.Y`?

> 2. Update the version number in `DESCRIPTION` to align with the Milestones created for this quarter 

Reviewer 2: ```suggestion
2. Update the version number in `DESCRIPTION` to align with the Milestones created for this quarter.
```

Because of this dependency structure, it is **critical** to release packages in the correct order. If a downstream package is released before its dependencies are updated, it may:

* Fail to build or check due to mismatched versions.
* Introduce inconsistencies in shared functions or data definitions.
* Break automated workflows (CI/CD, simulations, reporting pipelines).

To prevent these issues, follow the release sequence below:

1. **gsm.core** – foundational utilities used by nearly all other packages.
2. **gsm.mapping** – builds on `gsm.core` to provide study-specific mappings.
3. **gsm.kri** – depends on `gsm.core` and `gsm.mapping`; provides core risk indicator functionality.
4. **gsm.reporting** – depends on `gsm.core` and `gsm.kri`; underpins all reporting templates.
5. **gsm.datasim** – depends on `gsm.core` and `gsm.mapping` for generating simulated study data.
6. **gsm.qtl** – depends on `gsm.core` and `gsm.kri`.
7. **gsm.endpoints** – uses `gsm.mapping` and `gsm.core`. Also extends `gsm.kri` and `gsm.reporting` for endpoint-specific analyses.

> 8. **grail** – suggests `gsm.core`, `gsm.mapping`, `gsm.kri` and `gsm.reporting`.

Reviewer: Is `suggests` a synonym to `depends` in this case?

Reviewer 3: suggests is a separate section in the github DESCRIPTION file, and is typically used when the package functionality doesn't _require_ the package, but it is used in documentation/vignettes etc.

Reviewer: Thanks for clarification, totally forgot about it!

9. **grail.ado** - depends on `grail` for passing data to ADO.
10. **gsm.rrm** - suggests `grail` and `gsm.core`.
11. **gsm.template** – depends on `gsm.core`, `gsm.datasim` and `gsm.mapping`.
12. **gsm.app** – depends on `gsm.core` and `gsm.kri`. Suggests `gsm.mapping` and `gsm.reporting`.`
13. **gsm.ae** – depends on `gsm.app` and `gsm.mapping`
14. **gsm.qc** – depends on `gsm.core`, `gsm.mapping`, `gsm.kri` and `gsm.reporting`.

> 15. **gsm.utils** – optional utilities; No gsm dependencies at this time

Reviewer 2: ```suggestion
1. **gsm.core** – Foundational utilities used by nearly all other packages.
2. **gsm.mapping** – Builds on `gsm.core` to provide study-specific mappings.
3. **gsm.kri** – Depends on `gsm.core` and `gsm.mapping`; provides core risk indicator functionality.
4. **gsm.reporting** – Depends on `gsm.core` and `gsm.kri`; underpins all reporting templates.
5. **gsm.datasim** – Depends on `gsm.core` and `gsm.mapping` for generating simulated study data.
6. **gsm.qtl** – Depends on `gsm.core` and `gsm.kri`.
7. **gsm.endpoints** – Uses `gsm.mapping` and `gsm.core`. Also extends `gsm.kri` and `gsm.reporting` for endpoint-specific analyses.
8. **grail** – Suggests `gsm.core`, `gsm.mapping`, `gsm.kri` and `gsm.reporting`.
9. **grail.ado** - Depends on `grail` for passing data to ADO.
10. **gsm.rrm** - Suggests `grail` and `gsm.core`.
11. **gsm.template** – Depends on `gsm.core`, `gsm.datasim` and `gsm.mapping`.
12. **gsm.app** – Depends on `gsm.core` and `gsm.kri`. Suggests `gsm.mapping` and `gsm.reporting`.`
13. **gsm.ae** – Depends on `gsm.app` and `gsm.mapping`
14. **gsm.qc** – Depends on `gsm.core`, `gsm.mapping`, `gsm.kri` and `gsm.reporting`.
15. **gsm.utils** – Optional utilities. No gsm dependencies at this time
```

**Note:** Not every package requires a release in every cycle. If no changes have been made (and the version number/NEWS.md do not need updating), simply skip that package and continue with the next in the sequence.

By following this order, you ensure that every package is released with its latest compatible dependencies already available, minimizing risk of conflicts and downstream errors.

---
  
## Post-Release Steps
  
* Verify that all release tags are visible on GitHub.

> * Add qualification documentation to `r-qualification` repo, as required

Reviewer 2: ```suggestion
* Add qualification documentation to `r-qualification` repo, as required.
```

## Issues

Issues are the primary way to communicate what needs to be done and to track progress. Several issue templates are provided to help streamline this process, including: 

- [Requirements](https://github.com/Gilead-BioStats/gsm.core/issues/new?template=business_requirement.md): Use this template to create a User Requirement. Requirements are then assigned sub-issues using the issue types below. 
- [Bug](https://github.com/Gilead-BioStats/gsm.core/issues/new?template=bug_report.md): Bug reports for when something isn't working
- [Feature](https://github.com/Gilead-BioStats/gsm.core/issues/new?template=feature_request.md): User-facing Functionality
- [Technical Task](https://github.com/Gilead-BioStats/gsm.core/issues/new?template=qc_update.md): Non-user facing tasks, such as infrastructure updates or internal tooling improvements.
- [Documentation Task](https://github.com/Gilead-BioStats/gsm.core/issues/new?template=documentation.md): Improvements or additions to the documentation including function docs, readme updates and vignettes.

> The issue templates are maintained in {gsm.utils} and automatically updated in all gsm packages whenever updates are made. Note that suggestions or other input that might not warrant formal submission of an issue can be filed under [discussions](https://github.com/Gilead-BioStats/gsm.core/discussions), which can help facilitate discourse of specific use-cases or requests.

Reviewer 3: ```suggestion
The issue templates are maintained in {[gsm.utils](https://github.com/Gilead-BioStats/gsm.utils)} and automatically updated in all gsm packages whenever updates are made. Note that suggestions or other input that might not warrant formal submission of an issue can be filed under [discussions](https://github.com/Gilead-BioStats/gsm.core/discussions), which can help facilitate discourse of specific use-cases or requests.
```

## Road Map

The [gsm Roadmap](https://github.com/orgs/Gilead-BioStats/projects/41) GitHub Project is used to track issues and set priorities across projects. The gsm team generally batches work in to quarterly releases, although package releases can happen more frequently. Key cross-repo views in the project include: 

- Roadmap - a list of issues split out by quarter. Generally speaking, related functionality should be grouped in to a Requirement. Key columns include: 
   - Status - Issue status (e.g. To Do, In Progress, Done)
   - Repository/Milestone
   - Roadmap - Quarter where the requirement is planned for completion
   - Triage - has the requirement been approved for development work
   - Type - Bug, Feature, Task, Documentation
- Issues - a list of all open non-requirement issues, split by type.
- Releases - All open issues grouped by repo and sorted by milestone. 

# Development Workflow

> Development is done using R package development best practices (Add link to book) and Github Flow (Add link).

Reviewer 3: ```suggestion
Development is done using R package development best practices (See [R Packages (2e)](https://r-pkgs.org/)) and [Github Flow](https://docs.github.com/en/get-started/using-github/github-flow).
```
